Sources/BraveWallet/Crypto/Stores/AccountsStore.swift

```diff
@@ -0,0 +1,261 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import BraveCore
+import SwiftUI
+import Preferences
+
+struct AccountDetails: Equatable, Identifiable {
+  var id: String { account.id }
+  let account: BraveWallet.AccountInfo
+  let tokensWithBalance: [BraveWallet.BlockchainToken]
+  let totalBalanceFiat: String
+}
+
+class AccountsStore: ObservableObject, WalletObserverStore {
+  
+  /// Users primary accounts
+  @Published var primaryAccounts: [AccountDetails] = []
+  /// Users imported accounts
+  @Published var importedAccounts: [AccountDetails] = []
+  /// If we are loading prices or balances
+  @Published var isLoading: Bool = false
+  /// Users selected currency code
+  @Published private(set) var currencyCode: String = CurrencyCode.usd.code {
+    didSet {
+      currencyFormatter.currencyCode = currencyCode
+      guard oldValue != currencyCode else { return }
+      update()
+    }
+  }
+  
+  let currencyFormatter: NumberFormatter = .usdCurrencyFormatter
+
+  /// Cache of token balances for each account. [account.address: [token.id: balance]]
+  private var tokenBalancesCache: [String: [String: Double]] = [:]
+  /// Cache of prices for each token. The key is the token's `assetRatioId`.
+  private var pricesCache: [String: String] = [:]
+  
+  private let keyringService: BraveWalletKeyringService
+  private let rpcService: BraveWalletJsonRpcService
+  private let walletService: BraveWalletBraveWalletService
+  private let assetRatioService: BraveWalletAssetRatioService
+  private let userAssetManager: WalletUserAssetManagerType
+  
+  private var keyringServiceObserver: KeyringServiceObserver?
+  private var walletServiceObserver: WalletServiceObserver?
+
+  var isObserving: Bool {
+    keyringServiceObserver != nil
+  }
+  
+  init(
+    keyringService: BraveWalletKeyringService,
+    rpcService: BraveWalletJsonRpcService,
+    walletService: BraveWalletBraveWalletService,
+    assetRatioService: BraveWalletAssetRatioService,
+    userAssetManager: WalletUserAssetManagerType
+  ) {
+    self.keyringService = keyringService
+    self.rpcService = rpcService
+    self.walletService = walletService
+    self.assetRatioService = assetRatioService
+    self.userAssetManager = userAssetManager
+    self.setupObservers()
+  }
+  
+  func setupObservers() {
+    guard !isObserving else { return }
+    self.keyringServiceObserver = KeyringServiceObserver(
+      keyringService: keyringService,
+      _accountsChanged: { [weak self] in
+        self?.update()
+      }
+    )
+    self.walletServiceObserver = WalletServiceObserver(
+      walletService: walletService,
+      _onNetworkListChanged: { [weak self] in
+        self?.update()
+      }
+    )
+    Preferences.Wallet.showTestNetworks.observe(from: self)
+  }
+  
+  func tearDown() {
+    keyringServiceObserver = nil
```

Comment:
need to nil `walletServiceObserver` too

Reply:
Nice catch, fixed in https://github.com/brave/brave-ios/pull/8659/commits/6824b33977f8cd4806000e379946f52bb16ddee6